### DISCUSSION TITLE
queryClient.setQueryDefaults('user', {enabled:true}) not working?

### DISCUSSION THREAD

**Discussion by tamis-laan** [TanStack/query#3255] [Q&A]
I'm experimenting with authentication and I have two mutations useLogin and useLogout aswel as a query useUser which returns the user object. 

The user object should be undefined when not logged in and defined when loged in. Also the user object should not be fetched when logged out.

User query
``` javascript
export function useUser() {
	return useQuery('user', fetchUser, {enabled: false})
}
```

Login mutation
``` javascript
export function useLogin() {
	const queryClient = useQueryClient()
	return useMutation(fetchToken, {
		onSuccess: () => {
			queryClient.setQueryDefaults('user', {enabled:true})
			queryClient.invalidateQueries('user')
		}
	})
}
```

I would expect after using the useLogin hook that the useUser hook wil start fetching the user object as it has been set to `enabled:true` using the setQueryDefaults method. 

However this is not the case...

Is there something I'm not understanding here, or is this a bug?

**Comment by TkDodo** [comment]
a "default" is a value that is taken when you don't provide one. It's a bit like object merging:

```
{ ...defaultOptions, ...optionsFromUseQuery }
```

since you provide `enabled` in `useQuery`, that would translate to, in your example:

```
{ ...{ enabled: true }, ...{ enabled: false } }
```

which would yield `false`. So no, a default value can never override a concrete value set on `useQuery`.
